File: packages/py/cairn-p2p/src/cairn/pairing/payload.py
"""Pairing payload: data exchanged during pairing initiation."""

from __future__ import annotations

import time
from dataclasses import dataclass

import cbor2


@dataclass
class ConnectionHint:
    """A hint for how to reach a peer."""

    hint_type: str
    value: str


@dataclass
class PairingPayload:
    """The data exchanged during pairing initiation.

    CBOR key mapping: 0=peer_id, 1=nonce, 2=pake_credential,
    3=hints, 4=created_at, 5=expires_at.
    """

    peer_id: bytes  # 34-byte multihash
    nonce: bytes  # 16 bytes
    pake_credential: bytes
    connection_hints: list[ConnectionHint] | None
    created_at: int  # unix seconds
    expires_at: int  # unix seconds

    def is_expired(self, now_unix: int | None = None) -> bool:
        """Check whether this payload has expired."""
        if now_unix is None:
            now_unix = int(time.time())
        return now_unix > self.expires_at

    def to_cbor(self) -> bytes:
        """Serialize to CBOR with compact integer keys."""
        m: dict[int, object] = {
            0: self.peer_id,
            1: self.nonce,
            2: self.pake_credential,
        }
        if self.connection_hints:
            m[3] = [
                [h.hint_type, h.value]
                for h in self.connection_hints
            ]
        m[4] = self.created_at
        m[5] = self.expires_at
        return cbor2.dumps(m)

    @classmethod
    def from_cbor(cls, data: bytes) -> PairingPayload:
        """Deserialize from CBOR with compact integer keys."""
        m = cbor2.loads(data)
        if not isinstance(m, dict):
            raise ValueError("expected CBOR map")

        peer_id = bytes(m[0])
        nonce = bytes(m[1])
        if len(nonce) != 16:
            raise ValueError("nonce must be 16 bytes")
        pake_credential = bytes(m[2])

        hints = None
        if 3 in m:
            hints = [
                ConnectionHint(hint_type=h[0], value=h[1])
                for h in m[3]
            ]

        return cls(
            peer_id=peer_id,
            nonce=nonce,
            pake_credential=pake_credential,
            connection_hints=hints,
            created_at=m.get(4, 0),
            expires_at=m.get(5, 0),
        )

File: packages/ts/cairn-p2p/src/pairing/payload.ts
import { encode, decode } from 'cborg';
import { CairnError } from '../errors.js';

/** A hint for how to reach a peer. */
export interface ConnectionHint {
  hintType: string;
  value: string;
}

/**
 * The data exchanged during pairing initiation.
 * Contains everything a peer needs to bootstrap a connection and PAKE handshake.
 */
export interface PairingPayload {
  /** 32-byte Peer ID (SHA-256 hash of Ed25519 public key). */
  peerId: Uint8Array;
  /** 16-byte one-time nonce for this pairing attempt. */
  nonce: Uint8Array;
  /** PAKE credential (derived from mechanism-specific secret). */
  pakeCredential: Uint8Array;
  /** Optional connection hints (rendezvous addresses, etc.). */
  hints?: ConnectionHint[];
  /** Unix timestamp (seconds) when payload was created. */
  createdAt: number;
  /** Unix timestamp (seconds) when payload expires. */
  expiresAt: number;
}

/**
 * Serialize a PairingPayload to CBOR using compact integer keys.
 *
 * Key mapping (matches Rust): 0=peer_id, 1=nonce, 2=pake_credential, 3=hints, 4=created_at, 5=expires_at
 */
export function encodePairingPayload(payload: PairingPayload): Uint8Array {
  const map = new Map<number, unknown>();
  map.set(0, payload.peerId);
  map.set(1, payload.nonce);
  map.set(2, payload.pakeCredential);

  if (payload.hints && payload.hints.length > 0) {
    const hintArrays = payload.hints.map(h => [h.hintType, h.value]);
    map.set(3, hintArrays);
  }

  map.set(4, payload.createdAt);
  map.set(5, payload.expiresAt);

  return encode(map);
}

/**
 * Deserialize a PairingPayload from CBOR with compact integer keys.
 */
export function decodePairingPayload(data: Uint8Array): PairingPayload {
  try {
    const map = decode(data, { useMaps: true }) as Map<number, unknown>;

    const peerId = map.get(0) as Uint8Array;
    const nonce = map.get(1) as Uint8Array;
    const pakeCredential = map.get(2) as Uint8Array;

    if (!peerId || !nonce || !pakeCredential) {
      throw new CairnError('PAIRING', 'missing required fields in pairing payload');
    }

    let hints: ConnectionHint[] | undefined;
    const rawHints = map.get(3);
    if (rawHints && Array.isArray(rawHints)) {
      hints = rawHints.map((h: unknown) => {
        const arr = h as string[];
        return { hintType: arr[0], value: arr[1] };
      });
    }

    const createdAt = (map.get(4) as number) ?? 0;
    const expiresAt = (map.get(5) as number) ?? 0;

    return { peerId, nonce, pakeCredential, hints, createdAt, expiresAt };
  } catch (e) {
    if (e instanceof CairnError) throw e;
    throw new CairnError('PAIRING', `failed to decode pairing payload: ${e}`);
  }
}

/**
 * Check whether a payload has expired relative to the given unix timestamp (seconds).
 */
export function isPayloadExpired(payload: PairingPayload, nowUnix?: number): boolean {
  const now = nowUnix ?? Math.floor(Date.now() / 1000);
  return now > payload.expiresAt;
}

/**
 * Generate a 16-byte random nonce.
 */
export function generateNonce(): Uint8Array {
  return crypto.getRandomValues(new Uint8Array(16));
}
